Stripe Terminal SDK update 4.2.0

staskus opened the pull request:
<!-- Remember about a good descriptive title. -->

<!-- Id number of the GitHub issue this PR addresses. -->
<!-- Part of: # -> use this if your PR is one of many related to one issue -->

<!-- 
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
🛠 Need an AI review?  
Add a comment: `@coderabbitai review`
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
-->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

Updating [Stripe Terminal SDK update 4.2.0](https://github.com/stripe/stripe-terminal-ios/releases).

### Breaking Changes

SDK has had breaking changes in the API, some of which are affecting the Woo app:

1. A single connectReader. It also moves the delegate setting from the method to the configuration

> There is now a single [connectReader](https://stripe.dev/stripe-terminal-ios/docs/Classes/SCPTerminal.html#/c:objc(cs)SCPTerminal(im)connectReader:delegate:connectionConfig:completion:) call used for all connection methods. This replaces the previous methods: connectBluetoothReader, connectInternetReader, and connectLocalMobileReader.

2. A change in how disconnects are reported

In my testing, replacing `didReportUnexpectedReaderDisconnect` with `reader:didDisconnect:` works. However, in practice, this method is only triggered 1 minute after the reader connection is lost (e.g. reader is turned off). As I understand, it happens because due to automatic reconnection attempts in the background.

> Update: [Auto reconnect on unexpected disconnect](https://docs.stripe.com/terminal/payments/connect-reader?terminal-sdk-platform=ios&reader-type=tap-to-pay#2.-automatically-attempt-reconnection) is now enabled by default for mobile and Tap to Pay readers.
> Update: The method for handling reader disconnects has changed.
Removed terminal:didReportUnexpectedReaderDisconnect: from the SCPTerminalDelegate. Use [reader:didDisconnect:](https://stripe.dev/stripe-terminal-ios/docs/Protocols/SCPReaderDelegate.html#/c:objc(pl)SCPReaderDelegate(im)reader:didDisconnect:) to be informed of reader disconnects.

3. A few other renamings

Most notably, `appleBuiltIn` and `localMobile` is now consistently called `tapToPay`. I will make the same change in Woo codebase for consistency:

- https://github.com/woocommerce/woocommerce-ios/pull/15380

> [BluetoothReaderDelegate](https://stripe.dev/stripe-terminal-ios/3.9.0/Protocols/SCPBluetoothReaderDelegate.html) has been renamed to [MobileReaderDelegate](https://stripe.dev/stripe-terminal-ios/Protocols/SCPMobileReaderDelegate.html).

> Update: Renamed "local mobile" and "apple built in" to "Tap To Pay" in all SDK types, function names, and error codes to align with Stripe branding for this functionality.

4. Payment method type is now strongly typed

> Update: SCPPaymentIntentParameters and SCPSetupIntentParameters now keep payment method types as values of the SCPPaymentMethodType enum rather than strings.

### Other possibly relevant changes

1. New `discovering` state and enforcing a single discovery operation

We already have our own `discovering` state when starting the card reader process.

[We also have logic that uses locks](https://github.com/woocommerce/woocommerce-ios/blob/e514d68f295dafb8e7b9d6d0ba201dcd64edd228/Hardware/Hardware/CardReader/StripeCardReader/StripeCardReaderService.swift#L167) to make sure another discovery process is not starting or canceling. I'll test separately if this change allows us to remove [the locks developed in this PR.](https://github.com/woocommerce/woocommerce-ios/pull/4351)

> New: Added a new enum value discovering to [SCPConnectionStatus](https://stripe.dev/stripe-terminal-ios/docs/Enums/SCPConnectionStatus.html) to represent when discovery is running.
> Update: Subsequent calls to SCPTerminal's discoverReaders:delegate:completion: cancel all previously queued discovery operations. Only one discovery operation can run at any given time; all other discovery attempts will fail with SCPErrorCanceledDueToIntegrationError.

2. Changes in cancelation

I don't know if this is worth exploring. Have we been missing some flows where cancelation should be supported?

> Update: SCPTerminal's confirmPaymentIntent:completion, confirmSetupIntent:completion, and confirmRefund:completion operations now return SCPCancelable's that allow you to cancel the operation in certain scenarios.

I tested payment cancelation and it continues to work.

> Update: Calls to SCPTerminal's cancelPaymentIntent:completion or cancelSetupIntent:completion will now cancel ongoing operations related to the specified intent.

## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

Feel free to pick a few scenarios, especially the ones that were tested with a simulator (e.g. Interac payments). Also, one of the larger changes was done around reporting of unexpected disconnection, so it's good to test some these scenarios.

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

I tested IPP scenarios on iPhone 14 Pro (17.7) and POS on iPad Air M2 (18.2). I also used a simulated card reader for scenarios that weren't easily reproducible.

### IPP

| Tasks                                                     | UK   | US   | CA   |
|-----------------------------------------------------------|------|------|------|
| Connection from Order payment flow                        | ✅ | ✅ | ✅    |
| Connection from Payments menu                             | ✅   | ✅ | ✅  |
| Connection of a reader with a mandatory update            | ✅ | ✅ (simulator)  | ✅   |
| Connection of M2 reader                                   | N/A  | ✅  | N/A  |
| Connection of Chipper reader                               | N/A | ✅  | N/A  |
| Connection of WisePad 3 reader                            | ✅  | N/A  | ✅   |
| Connection from switched off state                         | ✅  | ✅  |  ✅   |
| Take a payment                         | ✅  | ✅ |  ✅    |
| Refund a payment                         | ✅ | ✅  | ✅   |
| Take payment after the in-line connection process         | ✅  | ✅ |  ✅ |
| Take payment with a previously-connected reader           | ✅  | ✅ | ✅ |
| Take TPP payment                                         | ✅   | ✅ | N/A   |
| Refund TTP payment                                | ✅  | ✅ | N/A   |
| Take card-inserted payment                                | ✅  | ✅ |  ✅    |
| Take Interac payment                                     | N/A | N/A  | ✅ (simulator)    |
| Refund Interac payment                                   | N/A | N/A  | ✅ (simulator)   |
| Reader update: error message when attempted with low battery | ✅ (simulator)  | ✅ (simulator) | ✅ (simulator)  |
| Reader update: successfully updates software and starts payment (in that flow) when completed | ✅  | ✅ (simulator)  | ✅   |

### POS

| Tasks                                                     | UK   | US   |
|-----------------------------------------------------------|------|------|
| Connection of M2 reader                                   | N/A  | ✅  |
| Connection of Chipper reader                               | N/A | ✅  |
| Connection of WisePad 3 reader                            | ✅ | N/A  |
| Connection from switched off state                         | ✅  | ✅  |
| Take payment after the in-line connection process         | ✅  | ✅ |
| Take payment with a previously-connected reader           | ✅ | ✅ |
| Reader update: error message when attempted with low battery | ☑️ (simulator)  | ✅ (simulator) |
| Reader update: successfully updates software and starts payment (in that flow) when completed | ☑️  | ✅ (simulator)  |
| Payment cancelations (going back, disconnecting. exiting POS) | ✅   | ✅   |

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

staskus committed:
Rename BluetoothReaderDelegate to MobileReaderDelegate

staskus committed:
Rename localMobile or appleBuiltIn SDK calls to tapToPay

staskus committed:
Use a single connectReader call instead of connectBluetoothReader or connectLocalMobileReader

staskus committed:
Update errors to reference tapToPay instead of appleBuiltIn

staskus committed:
Use PaymentMethodType instead of String

Updating based on SDK update:

SCPPaymentIntentParameters and SCPSetupIntentParameters now keep payment method types as values of the SCPPaymentMethodType enum rather than strings.

staskus committed:
Update StripeTerminalSDK to 4.2.0 in Podfile

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

staskus committed:
Replace terminalDidReportUnexpectedReaderDisconnect with readerDidDisconnect

In Stripe Terminal SDK, the method for handling reader disconnects has changed. Removed terminal:didReportUnexpectedReaderDisconnect: from the SCPTerminalDelegate. Use reader:didDisconnect: to be informed of reader disconnects.

staskus committed:
Update PaymentIntentParameters.swift

staskus committed:
Update RELEASE-NOTES.txt

jaclync commented:
👋 Apology for the delayed response, I started the review but haven't finished it and tested it yet. I will make sure to prioritize this Wednesday and finish the review by EOD Wednesday. Additional review(s) would also be helpful.

jaclync commented on Hardware/Hardware/CardReader/PaymentMethodType.swift:
```diff
@@ -0,0 +1,4 @@
+public enum PaymentMethodType {
```
nit: maybe we can add a comment potentially with a link to the Stripe doc, as it seems to mirror a subset of `StripeTerminal.PaymentMethodType`?

staskus replied:
Good idea. I will.

I borrowed the approach of `CardReaderType` and `CardReaderType+Stripe`, it's a similar type used in a similar way.

jaclync commented on RELEASE-NOTES.txt:
```diff
@@ -8,6 +8,7 @@
 - [internal] Assign `siteID` and `productID` to image product upload statuses. [https://github.com/woocommerce/woocommerce-ios/pull/15196]
 - [*] Payments: Improved payment views' adaptability to larger accessibility font sizes [https://github.com/woocommerce/woocommerce-ios/pull/15328].
 - [internal] Add site related properties to crowdsignal surveys. [https://github.com/woocommerce/woocommerce-ios/pull/15353]
+- [Internal] Update Stripe Terminal SDK to 4.2.0 [https://github.com/woocommerce/woocommerce-ios/pull/15377]
```
super-duper nit:
```suggestion
- [internal] Update Stripe Terminal SDK to 4.2.0 [https://github.com/woocommerce/woocommerce-ios/pull/15377]
```

jaclync commented on Hardware/Hardware/CardReader/StripeCardReader/StripeCardReaderService.swift:
```diff
@@ -960,29 +958,33 @@ extension StripeCardReaderService: BluetoothReaderDelegate {
     }
+
+    public func reader(_ reader: Reader, didDisconnect reason: DisconnectReason) {
+        connectedReadersSubject.send([])
```
Just wondering, does TTP delegate have a similar method to `didDisconnect` in the bluetooth reader delegate now that `TerminalDelegate` is removed? Or TTP doesn't really have a "disconnecting" state?

staskus replied:
Yes. 

Both `TapToPayReaderDelegate` and `MobileReaderDelegate` have `didDisconnect`. These delegates are both `SCPReaderDelegate`, which defines [didDisconnect](https://stripe.dev/stripe-terminal-ios/docs/Protocols/SCPReaderDelegate.html).

It means that when we set delegate to `self`, this method is shared by both TTP and Reader delegates.

`BluetoothConnectionConfigurationBuilder(delegate: self, locationId: locationId)`
`TapToPayConnectionConfigurationBuilder(delegate: self, locationId: locationId)`

jaclync replied:
Thanks for the explanation, makes sense now. I wasn't aware both delegates extend the same base delegate.

jaclync commented on Yosemite/Yosemite/Model/Payments/CardPresentPaymentsConfiguration.swift:
```diff
@@ -3,7 +3,7 @@ import WooFoundation
 public struct CardPresentPaymentsConfiguration: Equatable {
     public let countryCode: CountryCode
-    public let paymentMethods: [WCPayPaymentMethodType]
+    public let paymentMethods: [PaymentMethodType]
```
❓  Is `WCPayPaymentMethodType` still used? Is it only meant to be used for WooPayments by its name?

staskus replied:
Yes. It's used for wider `WooPayments` case. `WCPayPaymentMethodType` is used by `Networking` layer to represent payment charges. It contains more payment types than are supported by IPP. It looks like it's aprt of `WCPayCharge` object that is used at least for the refunds.

`PaymentMethodType` is meant to be used by `Hardware` layer and represents payment types that are supported by IPP. We only used `WCPayPaymentMethodType` here to later change this value to String. However, Stripe SDK now accepts an enum, so it makes sense to define one within `Hardware` and ask in `collectPayment` method.

jaclync reviewed (approved):
Thanks for taking on this non-trivial SDK update, and the detailed PR description and testing scenarios! LGTM :shipit: 

The tasks I tested (I have an Interac test card):

IPP

Tasks | US | CA
-- | -- | --
Connection from Order payment flow | ✅ |  
Connection from Payments menu | ✅ |  
Connection of a reader with a mandatory update | ✅ (simulator) |  
Connection of M2 reader | ✅ | N/A
Connection of Chipper reader | ✅ | N/A
Connection of WisePad 3 reader | N/A | ✅
Connection from switched off state | ✅ |  
Take a payment | ✅ |  
Refund a payment | ✅ |  
Take payment after the in-line connection process | ✅ |  
Take payment with a previously-connected reader | ✅ |  
Take TPP payment | ✅ | N/A
Refund TTP payment | ✅ | N/A
Take card-inserted payment | ✅ |  
Take Interac payment | N/A | ✅
Refund Interac payment | N/A | ✅
Reader update: error message when attempted with low battery | ✅ (simulator) |  
Reader update: successfully updates software and starts payment (in that flow) when completed | ✅ (simulator) | ✅ with WisePad 3, the update took about 5 minutes

POS

Tasks | US
-- | --
Connection of M2 reader | ✅
Connection of Chipper reader | ✅
Connection of WisePad 3 reader | N/A
Connection from switched off state | ✅
Take payment after the in-line connection process | ✅
Take payment with a previously-connected reader | ✅
Reader update: error message when attempted with low battery | ✅ (simulator)
Reader update: successfully updates software and starts payment (in that flow) when completed | ✅ (simulator)
Payment cancelations (going back, disconnecting. exiting POS) | ✅

staskus committed:
Update RELEASE-NOTES.txt

Co-authored-by: Jaclyn Chen <watertranquil@gmail.com>

staskus committed:
Merge branch 'trunk' into task/stripe-sdk-update-4.2.0

staskus committed:
Add PaymentMethodType comment

joshheald commented:
Nice one on this @staskus – I read through it all and did some general testing, the changes look great to me.

iamgabrielma commented:
Sorry I'm late to the party: I was planning to test this earlier as part of my Beta testing rotation but I had to set myself AFK unexpectedly. I ran multiple of the cases above for IPP and POS using M2,  as well as TTP cases and saw no issues 👍 

Thanks for handling this @staskus
